### PR TITLE
Enable machine parsable output with --format json

### LINKED ISSUE #280: Switch to machine-parsable output
### TL;DR

Currently, we rely on parsing the text output from the gcloud command to extract the URL and other outputs. This isn't guaranteed to be stable. We should instead use `--format=json` and parse the resulting output as JSON to extract desired values. This also would enable use to expose more outputs about the revision that we currently do not.

### Detailed design

```markdown
Append `--format=json` to all gcloud commands, parse output as JSON.
```

### Additional information

_No response_

### PR DESCRIPTION
**Pull request by verbanicm**
Enable `--format json` for all gcloud commands to enable machine parsable output

Resolves #280

### CODE REVIEW

**Comment by verbanicm** [comment]
Changes in `output-parser.ts` are trying to enable existing text parsing functionality. This could probably be made better by adding a `tag_url` output parameter, but would alter existing functionality.

File: src/deploy-cloudrun.ts [hunk 1]
```diff
@@ -31,144 +60,160 @@ export async function run(): Promise<void> {
...
-    const secrets = core.getInput('secrets'); // String of secrets KEY=VALUE,...
-    const region = core.getInput('region') || 'us-central1';
-    const source = core.getInput('source'); // Source directory
-    const suffix = core.getInput('suffix');
-    const tag = core.getInput('tag');
-    const noTraffic = core.getBooleanInput('no_traffic');
-    const revTraffic = core.getInput('revision_traffic');
-    const tagTraffic = core.getInput('tag_traffic');
-    const flags = core.getInput('flags');
+    const inputs: DeployCloudRunInputs = {
```
Comment (sethvargo) [inline src/deploy-cloudrun.ts:63]: This is a pretty big change that might be better split up into a separate PR. It's also a bit of a departure from how our other actions establish state. I'm also a little worried about the values all becoming mutable now.

File: src/output-parser.ts [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+  statusTags: string;
+  tags: string[];
+  urls: string[];
+}
+
+export function parseUpdateTrafficResponse(inputs: DeployCloudRunInputs, stdout: string) {
+  const outputJSON: UpdateTrafficItem[] = JSON.parse(stdout);
+
+  // Validate response
+  if (!outputJSON.length) {
```
Comment (sethvargo) [inline src/output-parser.ts:40]:
```suggestion
  if (!outputJSON?.length) {
```

or just 

```node
  if (!outputJSON) {
```

File: src/output-parser.ts [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+    throw new Error(`command response is empty: ${stdout}`);
+  }
+
+  // Default to service url
+  const responseItem = outputJSON[0];
+  let url = responseItem?.serviceUrl;
+
+  // Maintain current logic to use first tag URL if present
+  const tagUrls: string[] = outputJSON
```
Comment (sethvargo) [inline src/output-parser.ts:49]: I think `find` would be more efficient here (or an old-fashioned for-loop) since we only care about the first value.

File: src/output-parser.ts [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { DeployCloudRunInputs, DeployCloudRunOutputs } from './deploy-cloudrun';
+import { run_v1 } from 'googleapis';
+
+export interface UpdateTrafficItem {
```
Comment (sethvargo) [inline src/output-parser.ts:20]: Doc

File: src/output-parser.ts [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+  serviceUrl: string;
+  specPercent: string;
+  specTags: string;
+  statusPercent: string;
+  statusTags: string;
+  tags: string[];
+  urls: string[];
+}
+
+export function parseUpdateTrafficResponse(inputs: DeployCloudRunInputs, stdout: string) {
```
Comment (sethvargo) [inline src/output-parser.ts:36]: Doc. Does this need to be exported?

File: src/output-parser.ts [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+  if (tagUrls.length) {
+    url = tagUrls[0];
+  }
+
+  const outputs: DeployCloudRunOutputs = { url };
+
+  return outputs;
+}
+
+export function parseDeployResponse(inputs: DeployCloudRunInputs, stdout: string) {
```
Comment (sethvargo) [inline src/output-parser.ts:62]: Doc. Does this need to be exported?

File: src/output-parser.ts [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+  const outputJSON: run_v1.Schema$Service = JSON.parse(stdout);
+
+  // Validate response
+  if (!outputJSON?.status?.url) {
+    throw new Error(`invalid command response: ${stdout}`);
```
Comment (sethvargo) [inline src/output-parser.ts:67]: Probably want a more specific error (or separate checking if the JSON is empty)

File: tests/unit/output-parser.test.ts [hunk 8]
```diff
@@ -0,0 +1,463 @@
...
+  tag: '',
+  noTraffic: false,
+  revTraffic: '',
+  tagTraffic: '',
+  flags: '',
+};
+
+describe('#output-parser', function () {
+  describe('#parseUpdateTrafficResponse', function () {
+    it('correctly parses update traffic outputs', function () {
```
Comment (sethvargo) [inline tests/unit/output-parser.test.ts:43]: (Can be a follow-up). I've been advocating for table-driven tests: https://mochajs.org/#dynamically-generating-tests

Either the table-driven or closure-driven work pretty well. Example: https://github.com/google-github-actions/actions-utils/blob/main/tests/encoding.test.ts

**Review by sethvargo — CHANGES_REQUESTED** [review]

File: src/deploy-cloudrun.ts (hunk 1 above)
Comment (verbanicm) [reply src/deploy-cloudrun.ts:63]: Thought here was to enable future outputs depending on input variables (IE pass inputs object to parsing functions)

We can use Object.freeze or I can revert back and we can create an options parameter to the parsing functions.

Comment (sethvargo) [reply src/deploy-cloudrun.ts:63]: I'm not a huge fan of magic since it makes debugging more difficult. I think I'd rather revert this and have explicit inputs/outputs.

Comment (verbanicm) [reply src/deploy-cloudrun.ts:63]: Ack.

File: src/output-parser.ts (hunk 3 above)
Comment (verbanicm) [reply src/output-parser.ts:49]: Was thinking ahead of myself, ack.

File: src/output-parser.ts (hunk 6 above)
Comment (verbanicm) [reply src/output-parser.ts:62]: Imported by deploy-cloudrun.ts

File: tests/unit/output-parser.test.ts [hunk 9]
```diff
@@ -0,0 +1,498 @@
...
+  parseUpdateTrafficResponse,
+  parseDeployResponse,
+  ParseInputs,
+} from '../../src/output-parser';
+
+describe('#output-parser', () => {
+  describe('#parseUpdateTrafficResponse', () => {
+    const cases: {
+      name: string;
+      input: { stdout: string };
```
Comment (sethvargo) [inline tests/unit/output-parser.test.ts:29]: What's the value of having a nested structure here? Are you thinking of supporting `stderr` for tests too? If not, it might be simpler to denest this.

File: tests/unit/output-parser.test.ts [hunk 10]
```diff
@@ -0,0 +1,498 @@
...
+                "https://test-1---test-basic-yaml-4goqgbaxqq-uc.a.run.app"
+              ]
+            }
+          ]
+        `,
+        },
+        expected: 'https://my-tag-1---test-basic-yaml-4goqgbaxqq-uc.a.run.app',
+      },
+      {
+        name: 'handles empty stdout',
```
Comment (sethvargo) [inline tests/unit/output-parser.test.ts:211]: This is actually empty array. Can we also add a test for literally no output?

File: tests/unit/output-parser.test.ts [hunk 11]
```diff
@@ -0,0 +1,498 @@
...
+                ],
+                "url": "https://hello-4goqgbaxqq-uc.a.run.app"
+              }
+            }
+          `,
+        },
+        expected: 'https://test---hello-4goqgbaxqq-uc.a.run.app',
+      },
+      {
+        name: 'handles invalid stdout',
```
Comment (sethvargo) [inline tests/unit/output-parser.test.ts:475]: Can we add a test for the response being `[]` and for the response being the empty string?

File: tests/unit/output-parser.test.ts [hunk 12]
```diff
@@ -0,0 +1,498 @@
...
+      },
+      {
+        name: 'handles empty stdout',
+        error: true,
+        input: {
+          stdout: '[]',
+        },
+        expected: 'gcloud response is empty: []',
+      },
+    ];
```
Comment (sethvargo) [inline tests/unit/output-parser.test.ts:218]: Add a test for the response being empty object `{}`

**Review by verbanicm — COMMENTED** [review]
Addressed comments

File: src/deploy-cloudrun.ts [hunk 13]
```diff
@@ -19,10 +19,26 @@ import { getExecOutput } from '@actions/exec';
...
+import { parseDeployResponse, parseUpdateTrafficResponse } from './output-parser';
 
 export const GCLOUD_METRICS_ENV_VAR = 'CLOUDSDK_METRICS_ENVIRONMENT';
 export const GCLOUD_METRICS_LABEL = 'github-actions-deploy-cloudrun';
 
+/**
+ * DeployCloudRunOutputs are the common GitHub action outputs created by this action
+ */
+export interface DeployCloudRunOutputs {
+  url?: string | null | undefined;
```
Comment (sethvargo) [inline src/deploy-cloudrun.ts:31]: I think by using `url?: string`, you implicitly allow undefined too. So maybe shorten this to:

```
url?: string
```

File: src/deploy-cloudrun.ts [hunk 14]
```diff
@@ -193,25 +217,24 @@ export async function run(): Promise<void> {
...
     }
 
-    // gcloud outputs status information to stderr.
-    // TODO: update this to use JSON or YAML machine-readable output instead.
-    setUrlOutput(output.stdout + output.stderr);
+    // Map outputs by response type
+    const outputs: DeployCloudRunOutputs =
+      responseType === ResponseTypes.UPDATE_TRAFFIC
+        ? parseUpdateTrafficResponse(output.stdout)
+        : parseDeployResponse(output.stdout, { tag });
```
Comment (sethvargo) [inline src/deploy-cloudrun.ts:224]:
```suggestion
        : parseDeployResponse(output.stdout, { tag: tag });
```

I dislike this particular node magic.

File: src/output-parser.ts [hunk 15]
```diff
@@ -0,0 +1,126 @@
...
+
+    // Maintain current logic to use first tag URL if present
+    for (const item of outputJSON) {
+      if (item?.urls?.length) {
+        url = item.urls[0];
+        break;
+      }
+    }
+
+    const outputs: DeployCloudRunOutputs = { url };
```
Comment (sethvargo) [inline src/output-parser.ts:75]:
```suggestion
    const outputs: DeployCloudRunOutputs = { url: url };
```

File: src/output-parser.ts [hunk 16]
```diff
@@ -0,0 +1,126 @@
...
+ * into a common DeployCloudRunOutputs object
+ *
+ * @param stdout
+ * @returns DeployCloudRunOutputs
+ */
+export function parseUpdateTrafficResponse(stdout: string | undefined): DeployCloudRunOutputs {
+  try {
+    stdout = presence(stdout);
+    if (!stdout || stdout === '{}' || stdout === '[]') {
+      throw new Error(`invalid input received`);
```
Comment (sethvargo) [inline src/output-parser.ts:58]: Add some context this is from parse update traffic:

```js
`failed to parse update traffic response: invalid input received`
```

File: src/output-parser.ts [hunk 17]
```diff
@@ -0,0 +1,126 @@
...
+        break;
+      }
+    }
+
+    const outputs: DeployCloudRunOutputs = { url };
+
+    return outputs;
+  } catch (err) {
+    const msg = errorMessage(err);
+    throw new Error(`failed to parse deploy response: ${msg}, stdout: ${stdout}`);
```
Comment (sethvargo) [inline src/output-parser.ts:80]: Add some context this is from parse update traffic:

```js
`failed to parse update traffic response: ${msg}, stdout: ${stdout}`
```

File: src/output-parser.ts [hunk 18]
```diff
@@ -0,0 +1,126 @@
...
+ * @returns DeployCloudRunOutputs
+ */
+export function parseDeployResponse(
+  stdout: string | undefined,
+  inputs: ParseInputs | undefined,
+): DeployCloudRunOutputs {
+  try {
+    stdout = presence(stdout);
+    if (!stdout || stdout === '{}' || stdout === '[]') {
+      throw new Error(`invalid input received`);
```
Comment (sethvargo) [inline src/output-parser.ts:99]: Add some context this is from parse update traffic:

```js
`failed to parse deploy response: invalid input received`
```

**Review by sethvargo — APPROVED** [review]
I think you'll need to rebase. I made a few small suggestions inline, but otherwise LGTM.

File: src/output-parser.ts (hunk 16 above)
Comment (verbanicm) [reply src/output-parser.ts:58]: This is caught by outer try and wraps with contextual message

File: src/output-parser.ts (hunk 18 above)
Comment (verbanicm) [reply src/output-parser.ts:99]: This is caught by outer try and wraps with contextual message
